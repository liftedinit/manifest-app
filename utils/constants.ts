import { ExtendedMetadataSDKType } from './types';

export const MFX_TOKEN_DATA: Omit<ExtendedMetadataSDKType, 'balance' | 'totalSupply'> = {
  description: 'The native token of the Manifest Chain',
  denom_units: [
    { denom: 'umfx', exponent: 0, aliases: [] },
    { denom: 'mfx', exponent: 6, aliases: [] },
  ],
  base: 'umfx',
  display: 'mfx',
  name: 'Manifest',
  symbol: 'MFX',
  uri: '',
  uri_hash: '',
};

export const OSMOSIS_TOKEN_DATA: Omit<ExtendedMetadataSDKType, 'balance' | 'totalSupply'> = {
  description: 'The native token of the Osmosis Chain',
  denom_units: [
    { denom: 'uosmo', exponent: 0, aliases: [] },
    { denom: 'osmo', exponent: 6, aliases: [] },
  ],
  base: 'uosmo',
  display: 'osmo',
  name: 'Osmosis',
  symbol: 'OSMO',
  uri: '',
  uri_hash: '',
};

export const tokenExponents = [
  { exponent: 18, subdenom: 'atto', letter: 'a', description: 'Smallest unit, 10⁻¹⁸' },
  { exponent: 15, subdenom: 'femto', letter: 'f', description: '10⁻¹⁵' },
  { exponent: 12, subdenom: 'pico', letter: 'p', description: '10⁻¹²' },
  { exponent: 9, subdenom: 'nano', letter: 'n', description: '10⁻⁹' },
  { exponent: 6, subdenom: 'micro', letter: 'u', description: '10⁻⁶' },
  { exponent: 3, subdenom: 'milli', letter: 'm', description: '10⁻³' },
];
