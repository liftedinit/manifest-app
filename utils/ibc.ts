import { Asset, AssetList, IBCInfo } from '@chain-registry/types';

import { Coin } from '@liftedinit/manifestjs/dist/codegen/cosmos/base/v1beta1/coin';

import { shiftDigits } from './maths';
import {
  assets as manifestAssets,
  ibc as manifestIbc,
} from 'chain-registry/testnet/manifesttestnet';
import { assets as osmosisAssets, ibc as osmosisIbc } from 'chain-registry/testnet/osmosistestnet';
import { assets as axelarAssets, ibc as axelarIbc } from 'chain-registry/testnet/axelartestnet';

const assets: AssetList[] = [manifestAssets, osmosisAssets, axelarAssets];

export const truncateDenom = (denom: string) => {
  return denom.slice(0, 10) + '...' + denom.slice(-6);
};

const filterAssets = (chainName: string, assetList: AssetList[]): Asset[] => {
  return (
    assetList
      .find(({ chain_name }) => chain_name === chainName)
      ?.assets?.filter(({ type_asset }) => type_asset === 'ics20' || !type_asset) || []
  );
};

const getAllAssets = (chainName: string) => {
  const nativeAssets = filterAssets(chainName, assets);
  const ibcAssets = filterAssets(chainName, assets);

  return [...nativeAssets, ...ibcAssets];
};

export const denomToAsset = (chainName: string, denom: string) => {
  const allAssets = getAllAssets(chainName);

  // Only handle IBC hashes
  if (denom.startsWith('ibc/')) {
    // Find the asset that has this IBC hash as its base
    const asset = allAssets.find(asset => asset.base === denom);
    if (asset?.traces?.[0]?.counterparty?.base_denom) {
      // Return the original denom from the counterparty chain
      return {
        ...asset,
        base: asset.traces[0].counterparty.base_denom,
      };
    }
  }

  // Return original asset if not an IBC hash
  return allAssets.find(asset => asset.base === denom);
};

export const denomToExponent = (chainName: string, denom: string) => {
  const asset = denomToAsset(chainName, denom);
  const unit = asset?.denom_units.find(({ denom }) => denom === asset.display);
  return unit?.exponent || 6;
};

export const prettyBalance = (chainName: string, balance: Coin) => {
  const { amount, denom } = balance;
  const asset = denomToAsset(chainName, denom);
  const symbol = asset?.symbol || truncateDenom(denom);
  const exponent = denomToExponent(chainName, denom);
  const displayAmount = shiftDigits(amount, -exponent);
  const logoUrl = Object.values(asset?.logo_URIs || {}).find(url => url);

  return { denom, symbol, amount, displayAmount, logoUrl, exponent };
};

export type PrettyBalance = ReturnType<typeof prettyBalance>;

const ibcData: IBCInfo[] = [...manifestIbc, ...osmosisIbc, ...axelarIbc];

export const getIbcInfo = (fromChainName: string, toChainName: string) => {
  let flipped = false;

  let ibcInfo = ibcData.find(
    i => i.chain_1.chain_name === fromChainName && i.chain_2.chain_name === toChainName
  );

  if (!ibcInfo) {
    ibcInfo = ibcData.find(
      i => i.chain_1.chain_name === toChainName && i.chain_2.chain_name === fromChainName
    );
    flipped = true;
  }

  if (!ibcInfo) {
    throw new Error('cannot find IBC info');
  }

  const key = flipped ? 'chain_2' : 'chain_1';
  const source_port = ibcInfo.channels[0][key].port_id;
  const source_channel = ibcInfo.channels[0][key].channel_id;

  return { source_port, source_channel };
};

export const getIbcDenom = (chainName: string, denom: string) => {
  const allAssets = getAllAssets(chainName);

  // Find the asset that has this denom as its counterparty base_denom
  const ibcAsset = allAssets.find(asset => asset.traces?.[0]?.counterparty?.base_denom === denom);

  // Return the IBC hash (base) if found
  return ibcAsset?.base;
};

export const normalizeIBCDenom = (chainName: string, denom: string) => {
  const asset = denomToAsset(chainName, denom);
  if (asset) {
    return {
      denom: asset.base,
    };
  }
  return { denom };
};

export type ResolvedIBCDenom = ReturnType<typeof normalizeIBCDenom>;
