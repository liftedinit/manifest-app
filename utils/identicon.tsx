/* eslint-disable @next/next/no-img-element */
import CryptoJS from 'crypto-js';
import Identicon, { IdenticonOptions } from 'identicon.js';
import React, { useEffect, useState } from 'react';

const COLORS: [number, number, number, number][] = [
  [56, 12, 197, 255],
  [160, 135, 255, 255],
  [88, 218, 210, 255],
  [225, 225, 249, 255],
];

export function identiconFromWalletAddress(walletAddress: string, size?: number) {
  const hash = CryptoJS.SHA256(walletAddress).toString(CryptoJS.enc.Hex);
  const colorIndex = parseInt(hash.charAt(hash.length - 1), 16) % COLORS.length;

  const options: IdenticonOptions = {
    foreground: COLORS[colorIndex],
    background: [255, 255, 255, 0],
    margin: 0.21,
    size: size ?? 700,
    format: 'svg',
  };

  return new Identicon(hash, options);
}

export const ProfileAvatar = ({
  walletAddress,
  size,
  withBackground = true,
}: {
  walletAddress?: string;
  size?: number;
  withBackground?: boolean;
}) => {
  const [avatarSrc, setAvatarSrc] = useState('');

  useEffect(() => {
    if (walletAddress) {
      const identicon = identiconFromWalletAddress(walletAddress, size);
      setAvatarSrc(`data:image/svg+xml;base64,${identicon.toString()}`);
    }
  }, [walletAddress, size]);

  const imageSize = size ? `${size}px` : '44px';

  return (
    <img
      src={avatarSrc}
      alt="Profile Avatar"
      className={`rounded-md ${withBackground ? 'dark:bg-[#FFFFFF0F] bg-[#0000000A]' : ''}`}
      style={{ height: imageSize, width: imageSize }}
    />
  );
};
