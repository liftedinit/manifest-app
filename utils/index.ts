export * from './voting';
export * from './staking';
export * from './maths';
export * from './logos';
export * from './ibc';
export * from './string';
export * from './identicon';
export * from './profanityFilter';
export * from './yupExtensions';
export * from './transactionUtils';
export * from './types';
export * from './constants';
