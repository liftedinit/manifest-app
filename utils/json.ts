export function isJsonString(str: string): boolean {
  try {
    const parsed = JSON.parse(str);
    return parsed !== null && (typeof parsed === 'object' || Array.isArray(parsed));
  } catch (e) {
    return false;
  }
}
