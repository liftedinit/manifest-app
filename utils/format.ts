import { MetadataSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/bank';

import env from '@/config/env';
import { shiftDigits } from '@/utils/maths';

import { denomToAsset } from './ibc';

/**
 * Format configuration for large numbers
 * [threshold value, suffix string, maximum fraction digits]
 * Ordered from largest to smallest threshold
 */
const SUFFIXES: [number, string, number][] = [
  [1e24, '_', 0], // Special case. No suffix for >= 1e24, just scientific notation.
  [1e18, 'QT', 2],
  [1e15, 'Q', 2],
  [1e12, 'T', 2],
  [1e9, 'B', 2],
  [1e6, 'M', 2],
  [0, '', 6],
  // Cannot be negative.
];

/**
 * Format a large number to a human-readable string.
 * @param num The number to format.
 */
export function formatLargeNumber(num: number): string {
  if (!Number.isFinite(num)) return 'Invalid number';
  if (num <= 0) return '0';

  if (num >= SUFFIXES[0][0]) {
    return `${num.toExponential(6).replace(/\.?0*e\+?/, 'e')}`;
  }

  for (const [value, suffix, maximumFractionDigits] of SUFFIXES) {
    if (num >= value) {
      let s = (suffix ? num / value : num).toLocaleString(undefined, {
        maximumFractionDigits,
      });

      return `${s}${suffix}`;
    }
  }

  return num.toLocaleString();
}

export function formatDenom(denom: string): string {
  const assetInfo = denomToAsset(env.chain, denom);

  // Fallback to cleaning the denom if no assetInfo
  let cleanDenom = denom.replace(/^factory\/[^/]+\//, '');

  // Skip cleaning for IBC denoms as they should be resolved via assetInfo
  if (cleanDenom.startsWith('ibc/')) {
    cleanDenom = assetInfo?.display.toUpperCase() ?? cleanDenom;
  } else if (cleanDenom.startsWith('u')) {
    cleanDenom = cleanDenom.slice(1).toUpperCase();
  }

  return cleanDenom;
}

export function formatAmount(amount: string, denom: string, metadata?: MetadataSDKType[]) {
  const meta = metadata?.find(m => m.base === denom);
  const exponent = Number(meta?.denom_units[1]?.exponent) || 6;
  return Number(shiftDigits(amount, -exponent));
}
