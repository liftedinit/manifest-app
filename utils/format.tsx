import { MetadataSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/bank';
import { shiftDigits } from '@/utils/maths';
import { denomToAsset } from './ibc';
import { DenomDisplay, DenomVerifiedBadge } from '@/components/factory/components/DenomDisplay';
import env from '@/config/env';
import { ReactNode } from 'react';

export function formatLargeNumber(num: number): string {
  if (!Number.isFinite(num)) return 'Invalid number';
  if (num === 0) return '0';

  const quintillion = 1e18;
  const quadrillion = 1e15;
  const trillion = 1e12;
  const billion = 1e9;
  const million = 1e6;

  if (num < million) {
    return num.toString();
  }

  if (num >= quintillion) {
    return `${(num / quintillion).toFixed(2)}QT`;
  } else if (num >= quadrillion) {
    return `${(num / quadrillion).toFixed(2)}Q`;
  } else if (num >= trillion) {
    return `${(num / trillion).toFixed(2)}T`;
  } else if (num >= billion) {
    return `${(num / billion).toFixed(2)}B`;
  } else if (num >= million) {
    return `${(num / million).toFixed(2)}M`;
  }
  return num.toFixed(6);
}

/**
 * Format a denom to a display name react node with a verified badge.
 * @param denom The denom to format.
 * @param small Whether to use a smaller badge.
 */
export function formatDenomWithBadge(denom: string, small?: boolean): ReactNode {
  const cleanDenom = formatDenom(denom);
  const classes = `${small ? 'w-3' : 'w-5'} mx-1 inline relative bottom-1 text-primary`;

  return (
    <>
      {cleanDenom} <DenomVerifiedBadge base={denom} className={classes} />
    </>
  );
}

export function formatDenom(denom: string): string {
  const assetInfo = denomToAsset(env.chain, denom);

  // Fallback to cleaning the denom if no assetInfo
  let cleanDenom = denom.replace(/^factory\/[^/]+\//, '');

  // Skip cleaning for IBC denoms as they should be resolved via assetInfo
  if (cleanDenom.startsWith('ibc/')) {
    cleanDenom = assetInfo?.display.toUpperCase() ?? '';
  } else if (cleanDenom.startsWith('u')) {
    cleanDenom = cleanDenom.slice(1).toUpperCase();
  }

  return cleanDenom;
}

export function formatAmount(amount: string, denom: string, metadata?: MetadataSDKType[]) {
  const meta = metadata?.find(m => m.base === denom);
  const exponent = Number(meta?.denom_units[1]?.exponent) || 6;
  return Number(shiftDigits(amount, -exponent));
}
