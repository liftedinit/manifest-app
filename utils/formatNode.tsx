/**
 * A collection of functions to format data, returning react nodes.
 * This file is intended as a TSX addendum to `./format.ts`.
 */
import { ReactNode } from 'react';
import { DenomVerifiedBadge } from '@/components';
import { formatDenom } from '@/utils/format';

/**
 * Format a denom to a display name react node with a verified badge.
 * @param denom The denom to format.
 * @param small Whether to use a smaller badge.
 */
export function formatDenomWithBadge(denom: string, small?: boolean): ReactNode {
  const cleanDenom = formatDenom(denom);
  const classes = `${small ? 'w-3' : 'w-5'} inline relative bottom-1 text-primary`;

  return (
    <>
      {cleanDenom} <DenomVerifiedBadge base={denom} className={classes} />
    </>
  );
}
