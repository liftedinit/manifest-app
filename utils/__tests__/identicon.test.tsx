import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'bun:test';

import { formatComponent } from '@/tests';
import { ProfileAvatar } from '@/utils';

describe('ProfileAvatar', () => {
  test('renders correctly', () => {
    const mockup = render(
      <ProfileAvatar walletAddress="manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf" />
    );
    expect(screen.getByAltText('Profile Avatar')).toBeInTheDocument();

    expect(formatComponent(mockup.asFragment())).toMatchSnapshot();
  });
});
