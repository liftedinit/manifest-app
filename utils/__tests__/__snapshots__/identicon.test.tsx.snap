// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileAvatar renders correctly 1`] = `
"<DocumentFragment>
  <img
    alt="Profile Avatar"
    class="rounded-md dark:bg-[#FFFFFF0F] bg-[#0000000A]"
    src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc3MDAnIGhlaWdodD0nNzAwJyBzdHlsZT0nYmFja2dyb3VuZC1jb2xvcjpyZ2JhKDI1NSwyNTUsMjU1LDApOyc+PGcgc3R5bGU9J2ZpbGw6cmdiYSgyMjUsMjI1LDI0OSwxKTsgc3Ryb2tlOnJnYmEoMjI1LDIyNSwyNDksMSk7IHN0cm9rZS13aWR0aDozLjU7Jz48cmVjdCAgeD0nMzA5JyB5PScyMjgnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PSczMDknIHk9JzM5MCcgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PHJlY3QgIHg9JzMwOScgeT0nNDcxJyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMjI4JyB5PScxNDcnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PSczOTAnIHk9JzE0Nycgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PHJlY3QgIHg9JzIyOCcgeT0nMjI4JyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMzkwJyB5PScyMjgnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PScyMjgnIHk9JzQ3MScgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PHJlY3QgIHg9JzM5MCcgeT0nNDcxJyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMTQ3JyB5PSc0NzEnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PSc0NzEnIHk9JzQ3MScgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PC9nPjwvc3ZnPg=="
    style="height: 44px; width: 44px;"
  />
</DocumentFragment>"
`;
